internal/bridge/crunchybridgecluster/postgres.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/crunchydata/postgres-operator/internal/naming"
+	"github.com/crunchydata/postgres-operator/pkg/apis/postgres-operator.crunchydata.com/v1beta1"
+)
+
+// generatePostgresRoleSecret returns a Secret containing a password and
+// connection details for the appropriate database.
+func (r *CrunchyBridgeClusterReconciler) generatePostgresRoleSecret(
+	cluster *v1beta1.CrunchyBridgeCluster, roleSpec *v1beta1.CrunchyBridgeClusterRoleSpec,
+	clusterRole *bridge.ClusterRoleApiResource,
+) (*corev1.Secret, error) {
+	roleName := roleSpec.Name
+	secretName := roleSpec.SecretName
+	intent := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{
+		Namespace: cluster.Namespace,
+		Name:      secretName,
+	}}
+	intent.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Secret"))
+	initialize.StringMap(&intent.StringData)
+
+	intent.StringData["name"] = clusterRole.Name
+	intent.StringData["password"] = clusterRole.Password
+	intent.StringData["uri"] = clusterRole.URI
+
+	intent.Annotations = cluster.Spec.Metadata.GetAnnotationsOrNil()
+	intent.Labels = naming.Merge(
+		cluster.Spec.Metadata.GetLabelsOrNil(),
+		map[string]string{
+			naming.LabelCluster: cluster.Name,
+			naming.LabelRole:    naming.RoleCrunchyBridgeClusterPostgresRole,
+			naming.LabelCrunchyBridgeClusterPostgresRole: roleName,
+		})
+
+	err := errors.WithStack(r.setControllerReference(cluster, intent))
+
+	return intent, err
+}
+
+// reconcilePostgresRoles writes the objects necessary to manage roles and their
+// passwords in PostgreSQL.
+func (r *CrunchyBridgeClusterReconciler) reconcilePostgresRoles(
+	ctx context.Context, apiKey string, cluster *v1beta1.CrunchyBridgeCluster,
+) error {
+	_, _, err := r.reconcilePostgresRoleSecrets(ctx, apiKey, cluster)
+
+	// TODO: If we ever add a PgAdmin feature to CrunchyBridgeCluster, we will
+	// want to add the role credentials to PgAdmin here
+
+	return err
+}
+
+func (r *CrunchyBridgeClusterReconciler) reconcilePostgresRoleSecrets(
+	ctx context.Context, apiKey string, cluster *v1beta1.CrunchyBridgeCluster,
+) (
+	[]*v1beta1.CrunchyBridgeClusterRoleSpec, map[string]*corev1.Secret, error,
+) {
+	log := ctrl.LoggerFrom(ctx)
+	specRoles := cluster.Spec.Roles
+
+	// Index role specifications by PostgreSQL role name.
+	roleSpecs := make(map[string]*v1beta1.CrunchyBridgeClusterRoleSpec, len(specRoles))
+	for i := range specRoles {
+		roleSpecs[specRoles[i].Name] = specRoles[i]
+	}
+
+	// Gather existing role secrets
+	secrets := &corev1.SecretList{}
+	selector, err := naming.AsSelector(naming.CrunchyBridgeClusterPostgresRoles(cluster.Name))
+	if err == nil {
+		err = errors.WithStack(
+			r.Client.List(ctx, secrets,
+				client.InNamespace(cluster.Namespace),
+				client.MatchingLabelsSelector{Selector: selector},
+			))
+	}
+
+	// Index secrets by PostgreSQL role name and delete any that are not in the
+	// cluster spec.
+	roleSecrets := make(map[string]*corev1.Secret, len(secrets.Items))
```

Comment:
I'm curious about the use of this map of secrets. 

* You use it to store a secret if it exists as a K8s object and should exist (line 114)
* You use it to store a secret before it gets created (line 132)
* And then you use it to return all the secrets (line 142)

But when it gets returned, it might have secrets that were generated but not created, right? Like if there's an error on creation, it's still in the map.

Or, taking a step back, since we're not using that list of secrets that gets returned, maybe we don't need it? 

Like: 
1. we iterate over the list of secrets returned; 
2. if there's a secret in that list that doesn't match the spec.roles, we can delete it; 
3. if there's a secret that does match the spec.roles, we can skip it; 
4. and if there's spec.roles that don't exist yet as secrets, we create it.

Reply:
So, a lot of this was modeled after the user/secret creation for `postgresclusters`. You make a good point that we add the generated secret to the map before ensuring that it was successfully created in k8s which seems wrong. I can make a change to only add it to the map after we confirm that it is created in k8s (this should be changed for `postgrescluster` too).

Also good point that as of right now, we don't really care what the map contains since we don't use it, although that could change in the future if we add pgAdmin support or something similar.

In your numbered outline, I don't think we want to do #3 (skip a secret if it's already there) because we want to refresh it in case the password/URI string has changed.

Reply:
Spoke to Chris about this... The convention is that if there is an error, we don't pay attention to the other return values. So it essentially doesn't matter that the map has a secret that wasn't successfully created in k8s because the failed k8s creation returns an error that negates everything else.